# **Jobsheet 13 - Artificial Neural Network (ANN)**

## **Praktikum 1 - JST sederhana (2 layer) dengan forward pass dan backpropagation manual**

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.26845216898128865
Epoch 1000, Loss: 0.23743952738078528
Epoch 2000, Loss: 0.18136189805229946
Epoch 3000, Loss: 0.14793584924684955
Epoch 4000, Loss: 0.13742788936747824
Epoch 5000, Loss: 0.1331412007220042
Epoch 6000, Loss: 0.13094205807660247
Epoch 7000, Loss: 0.12963513273465638
Epoch 8000, Loss: 0.1287792320890651
Epoch 9000, Loss: 0.12817938561855696
Prediksi:
[[0.0418    ]
 [0.49833313]
 [0.94833704]
 [0.50484248]]


### **Tugas 1**

1. Mengubah jumlah neuron hiden layer menjadi 3

In [ ]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.3068925545283221
Epoch 1000, Loss: 0.1940046599023882
Epoch 2000, Loss: 0.13501645305005108
Epoch 3000, Loss: 0.03944073939866964
Epoch 4000, Loss: 0.014921874020480876
Epoch 5000, Loss: 0.008219931925919046
Epoch 6000, Loss: 0.00545620174537408
Epoch 7000, Loss: 0.0040126499614300945
Epoch 8000, Loss: 0.00314365034475404
Epoch 9000, Loss: 0.002569605387275079
Prediksi:
[[0.02486432]
 [0.95593289]
 [0.95035328]
 [0.0603007 ]]


2. Membandingkan hasil loss dengan konfigurasi awal.

Berdasarkan hasil percobaan, dengan menambah jumlah neuron pada hidden layer dari 2 menjadi 3 mampu meningkatkan kemampuan belajar model. Pada konfigurasi awal (2 neuron), model gagal mempelajari pola XOR sepenuhnya dan terjebak dalam kondisi stuck, yang ditandai dengan nilai loss yang tertahan tinggi (0.128) dan prediksi angka yang ragu-ragu di sekitar 0.5. Sebaliknya, dengan menggunakan 3 neuron, jaringan memiliki kapasitas yang lebih besar dan fleksibel untuk memisahkan data. Hal ini membuat proses belajar menjadi lancar, menghasilkan loss yang turun drastis hingga mendekati nol (0.002), serta memberikan hasil prediksi yang sangat tegas dan akurat sesuai target.

3. Menambahkan fungsi aktivasi ReLU dan bandingkan hasil

In [ ]:

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter (Tetap menggunakan 3 neuron hidden agar adil dengan percobaan sebelumnya)
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

np.random.seed(42) # Seed sama agar perbandingan fair
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# --- DEFINISI FUNGSI AKTIVASI BARU ---
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

def relu(x):
    return np.maximum(0, x)

def relu_derivative(x):
    # Turunan ReLU: 1 jika x > 0, selain itu 0
    return (x > 0).astype(float)

# Training
for epoch in range(10000):
    # --- FORWARD PASS ---
    z1 = np.dot(X, W1) + b1
    a1 = relu(z1)       # UBAH: Hidden layer sekarang pakai ReLU

    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)    # Output layer tetap Sigmoid

    # Hitung error
    error = y - a2

    # --- BACKPROPAGATION ---
    # Layer Output (Tetap Sigmoid)
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    # Layer Hidden (Sekarang ReLU)
    # Perhatikan: Turunan ReLU menggunakan z1 (input sebelum aktivasi)
    d_a1 = np.dot(d_a2, W2.T) * relu_derivative(z1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss:.6f}")

# Output akhir
print("\nPrediksi Akhir (ReLU di Hidden Layer):")
print(a2)

Epoch 0, Loss: 0.327478
Epoch 1000, Loss: 0.012797
Epoch 2000, Loss: 0.003283
Epoch 3000, Loss: 0.001731
Epoch 4000, Loss: 0.001147
Epoch 5000, Loss: 0.000846
Epoch 6000, Loss: 0.000666
Epoch 7000, Loss: 0.000547
Epoch 8000, Loss: 0.000462
Epoch 9000, Loss: 0.000400

Prediksi Akhir (ReLU di Hidden Layer):
[[0.02970009]
 [0.98605912]
 [0.98605934]
 [0.01163533]]
